src/voter.rs

```diff
@@ -0,0 +1,614 @@
+
+		let last_round_state = self.last_round_state.get().clone();
+		self.prevote(&last_round_state)?;
+		self.precommit(&last_round_state)?;
+
+		try_ready!(self.outgoing.poll());
+
+		if self.votes.completable() {
+			Ok(Async::Ready(()))
+		} else {
+			Ok(Async::NotReady)
+		}
+	}
+
+	fn prevote(&mut self, last_round_state: &RoundState<H>) -> Result<(), E::Error> {
+		match self.state.take() {
+			Some(State::Start(mut prevote_timer, precommit_timer)) => {
+				let should_prevote = match prevote_timer.poll() {
+					Err(e) => return Err(e),
+					Ok(Async::Ready(())) => true,
+					Ok(Async::NotReady) => self.votes.completable(),
+				};
+
+				if should_prevote {
+					if let Some(prevote) = self.construct_prevote(last_round_state)? {
+						debug!(target: "afg", "Casting prevote for round {}", self.votes.number());
+						self.outgoing.push(Message::Prevote(prevote));
+					}
+					self.state = Some(State::Prevoted(precommit_timer));
+				} else {
+					self.state = Some(State::Start(prevote_timer, precommit_timer));
+				}
+			}
+			x => { self.state = x; }
+		}
+
+		Ok(())
+	}
+
+	fn precommit(&mut self, last_round_state: &RoundState<H>) -> Result<(), E::Error> {
+		match self.state.take() {
+			Some(State::Prevoted(mut precommit_timer)) => {
+				let last_round_estimate = last_round_state.estimate.clone()
+					.expect("Rounds only started when prior round completable; qed");
+
+				let should_precommit = {
+					// we wait for the last round's estimate to be equal to or
+					// the ancestor of the current round's p-Ghost before precommitting.
+					self.votes.state().prevote_ghost.as_ref().map_or(false, |p_g| {
+						p_g == &last_round_estimate ||
+							match self.env.ancestry(last_round_estimate.0, p_g.0.clone()) {
+								Ok(_) => true,
+								Err(::Error::NotDescendent) => false,
+							}
+					})
+				} && match precommit_timer.poll() {
+					Err(e) => return Err(e),
+					Ok(Async::Ready(())) => true,
+					Ok(Async::NotReady) => self.votes.completable(),
+				};
+
+				if should_precommit {
+					debug!(target: "afg", "Casting precommit for round {}", self.votes.number());
+					let precommit = self.construct_precommit();
+					self.outgoing.push(Message::Precommit(precommit));
+					self.state = Some(State::Precommitted);
+				} else {
+					self.state = Some(State::Prevoted(precommit_timer));
+				}
+			}
+			x => { self.state = x; }
+		}
+
+		Ok(())
+	}
+
+	// construct a prevote message based on local state.
+	fn construct_prevote(&self, last_round_state: &RoundState<H>) -> Result<Option<Prevote<H>>, E::Error> {
+		let last_round_estimate = last_round_state.estimate.clone()
+			.expect("Rounds only started when prior round completable; qed");
+
+		let find_descendent_of = match self.primary_block {
+			None => {
+				// vote for best chain containing prior round-estimate.
+				last_round_estimate.0
+			}
+			Some(ref primary_block) => {
+				// we will vote for the best chain containing `p_hash` iff
```

Comment:
typo `iff`

Reply:
no, iff is shorthand for "if and only if"